library/tinkerpop/src/main/java/uk/gov/gchq/gaffer/tinkerpop/generator/GafferEntityGenerator.java

```diff
@@ -16,33 +16,29 @@
 
 package uk.gov.gchq.gaffer.tinkerpop.generator;
 
-import org.apache.tinkerpop.gremlin.structure.Property;
-import org.apache.tinkerpop.gremlin.structure.VertexProperty;
-
 import uk.gov.gchq.gaffer.data.element.Entity;
 import uk.gov.gchq.gaffer.data.generator.OneToOneElementGenerator;
 import uk.gov.gchq.gaffer.tinkerpop.GafferPopVertex;
 
-import java.util.Iterator;
-
 public class GafferEntityGenerator implements OneToOneElementGenerator<GafferPopVertex> {
     @Override
     public Entity _apply(final GafferPopVertex vertex) {
+        if (vertex == null) {
+            throw new IllegalArgumentException("Unable to convert a null GafferPopVertex Object");
+        }
+
         final Entity entity = new Entity(vertex.label(), vertex.id());
-        final Iterator<VertexProperty<Object>> vertPropItr = vertex.properties();
-        while (vertPropItr.hasNext()) {
-            final VertexProperty<Object> vertProp = vertPropItr.next();
-            if (null != vertProp.key()) {
+        // Tinkerpop allows nested properties under a key for Gaffer we need to flatten these so only one property per key
+        vertex.properties().forEachRemaining(vertProp -> {
+            if (vertProp.key() != null) {
                 entity.putProperty(vertProp.key(), vertProp.value());
             }
-            final Iterator<Property<Object>> propItr = vertProp.properties();
-            while (propItr.hasNext()) {
-                final Property<Object> prop = propItr.next();
-                if (null != prop.key()) {
+            vertProp.properties().forEachRemaining(prop -> {
+                if (prop.key() != null) {
                     entity.putProperty(prop.key(), prop.value());
                 }
```

Comment:
This is essentially repeated code. It could be reduced by defining a lambda earlier on. Though I'm not sure if this is worth it as it would only save a few lines of code. Might be more readable though?

Reply:
Not sure its possible with one lambda as you need to loop over the first iterable before you have access to the nested iterable open to suggestions but simplest I can think of is remove the null check as its not adding anything (see other comment) and do something like:

```java
vertex.properties().forEachRemaining(vertProp -> {
    entity.putProperty(vertProp.key(), vertProp.value());
    vertProp.properties().forEachRemaining(prop -> entity.putProperty(prop.key(), prop.value()));
});
```

Reply:
I was thinking of a function as a variable which could be reused inside the existing lambdas. But this is a moot point if the null check is removed as that will condense these as you've shown.